File: backend/app/api/search_jobs.py
from fastapi import APIRouter, Depends, HTTPException, status
from fastapi.responses import FileResponse
from sqlalchemy.orm import Session
from typing import List
import os

from app.db.session import get_db
from app.models.user import User
from app.models.search_job import SearchJob as SearchJobModel
from app.models.paper import Paper as PaperModel
from app.schemas.search_job import SearchJob, SearchJobCreate, SearchJobUpdate, SearchJobList
from app.schemas.paper import Paper, PaperList
from app.api.auth import get_current_user
from app.tasks.scraping_tasks import run_search_job, resume_failed_job

router = APIRouter()


@router.post("/", response_model=SearchJob, status_code=status.HTTP_201_CREATED)
async def create_search_job(
    job_data: SearchJobCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Create a new search job.

    Args:
        job_data: Search job parameters
        current_user: Current authenticated user
        db: Database session

    Returns:
        Created search job
    """
    # Validate parameters
    if job_data.start_year and job_data.end_year:
        if job_data.start_year > job_data.end_year:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="start_year must be less than or equal to end_year"
            )

    if not job_data.keywords_include:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="At least one inclusion keyword is required"
        )

    # Create job
    job = SearchJobModel(
        user_id=current_user.id,
        name=job_data.name,
        keywords_include=job_data.keywords_include,
        keywords_exclude=job_data.keywords_exclude or [],
        semantic_criteria=job_data.semantic_criteria.dict() if job_data.semantic_criteria else None,
        start_year=job_data.start_year,
        end_year=job_data.end_year,
        status="pending"
    )

    db.add(job)
    db.commit()
    db.refresh(job)

    # Start the background task
    task = run_search_job.delay(str(job.id))
    job.celery_task_id = task.id
    db.commit()

    return job


@router.get("/", response_model=SearchJobList)
async def list_search_jobs(
    skip: int = 0,
    limit: int = 100,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    List all search jobs for the current user.

    Args:
        skip: Number of records to skip
        limit: Maximum number of records to return
        current_user: Current authenticated user
        db: Database session

    Returns:
        List of search jobs
    """
    query = db.query(SearchJobModel).filter(SearchJobModel.user_id == current_user.id)

    total = query.count()
    jobs = query.order_by(SearchJobModel.created_at.desc()).offset(skip).limit(limit).all()

    return SearchJobList(jobs=jobs, total=total)


@router.get("/{job_id}", response_model=SearchJob)
async def get_search_job(
    job_id: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Get a specific search job.

    Args:
        job_id: Job UUID
        current_user: Current authenticated user
        db: Database session

    Returns:
        Search job details
    """
    job = db.query(SearchJobModel).filter(
        SearchJobModel.id == job_id,
        SearchJobModel.user_id == current_user.id
    ).first()

    if not job:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Search job not found"
        )

    return job


@router.patch("/{job_id}", response_model=SearchJob)
async def update_search_job(
    job_id: str,
    job_data: SearchJobUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Update a search job.

    Args:
        job_id: Job UUID
        job_data: Fields to update
        current_user: Current authenticated user
        db: Database session

    Returns:
        Updated search job
    """
    job = db.query(SearchJobModel).filter(
        SearchJobModel.id == job_id,
        SearchJobModel.user_id == current_user.id
    ).first()

    if not job:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Search job not found"
        )

    # Update fields
    if job_data.name is not None:
        job.name = job_data.name

    if job_data.status is not None:
        # Validate status transition
        valid_statuses = ["pending", "running", "paused", "completed", "failed"]
        if job_data.status not in valid_statuses:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Invalid status. Must be one of: {valid_statuses}"
            )
        job.status = job_data.status

    db.commit()
    db.refresh(job)

    return job


@router.delete("/{job_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_search_job(
    job_id: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Delete a search job.

    Args:
        job_id: Job UUID
        current_user: Current authenticated user
        db: Database session
    """
    job = db.query(SearchJobModel).filter(
        SearchJobModel.id == job_id,
        SearchJobModel.user_id == current_user.id
    ).first()

    if not job:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Search job not found"
        )

    # Delete CSV file if exists
    if job.csv_file_path and os.path.exists(job.csv_file_path):
        os.remove(job.csv_file_path)

    db.delete(job)
    db.commit()


@router.post("/{job_id}/resume", response_model=SearchJob)
async def resume_search_job(
    job_id: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Resume a failed search job.

    Args:
        job_id: Job UUID
        current_user: Current authenticated user
        db: Database session

    Returns:
        Resumed search job
    """
    job = db.query(SearchJobModel).filter(
        SearchJobModel.id == job_id,
        SearchJobModel.user_id == current_user.id
    ).first()

    if not job:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Search job not found"
        )

    if job.status != "failed":
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Only failed jobs can be resumed"
        )

    # Start the resume task
    resume_failed_job.delay(str(job.id))

    db.refresh(job)
    return job


@router.get("/{job_id}/papers", response_model=PaperList)
async def get_job_papers(
    job_id: str,
    skip: int = 0,
    limit: int = 100,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Get papers extracted for a specific search job.

    Args:
        job_id: Job UUID
        skip: Number of records to skip
        limit: Maximum number of records to return
        current_user: Current authenticated user
        db: Database session

    Returns:
        List of papers
    """
    # Verify job belongs to current user
    job = db.query(SearchJobModel).filter(
        SearchJobModel.id == job_id,
        SearchJobModel.user_id == current_user.id
    ).first()

    if not job:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Search job not found"
        )

    # Get papers for this job
    query = db.query(PaperModel).filter(PaperModel.search_job_id == job_id)
    total = query.count()
    papers = query.order_by(PaperModel.created_at.desc()).offset(skip).limit(limit).all()

    return PaperList(papers=papers, total=total)


@router.get("/{job_id}/download")
async def download_results(
    job_id: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Download CSV results for a completed search job.

    Args:
        job_id: Job UUID
        current_user: Current authenticated user
        db: Database session

    Returns:
        CSV file download
    """
    job = db.query(SearchJobModel).filter(
        SearchJobModel.id == job_id,
        SearchJobModel.user_id == current_user.id
    ).first()

    if not job:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Search job not found"
        )

    if job.status != "completed":
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Job is not completed yet"
        )

    if not job.csv_file_path or not os.path.exists(job.csv_file_path):
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Results file not found"
        )

    filename = f"{job.name.replace(' ', '_')}_{job.id}.csv"

    return FileResponse(
        path=job.csv_file_path,
        media_type="text/csv",
        filename=filename
    )

File: backend-node/src/api/searchJobs.ts
import { Router, Response } from 'express';
import { SearchJob, Paper } from '../models';
import { authenticate, AuthRequest } from './middleware/auth';
import { addSearchJobToQueue } from '../tasks/queue';
import logger from '../core/logger';
import path from 'path';
import fs from 'fs';
import { config } from '../core/config';
import glob from 'glob';
import { promisify } from 'util';

const router = Router();
const globAsync = promisify(glob);

// POST /jobs - Create a new search job
router.post('/', authenticate, async (req: AuthRequest, res: Response) => {
  try {
    const {
      name,
      keywords_include,
      keywords_exclude,
      semantic_criteria,
      semantic_batch_mode,
      generate_latex,
      start_year,
      end_year,
      max_results,
    } = req.body;

    // Validate parameters
    if (start_year && end_year && start_year > end_year) {
      return res.status(400).json({ detail: 'start_year must be less than or equal to end_year' });
    }

    if (!keywords_include || keywords_include.length === 0) {
      return res.status(400).json({ detail: 'At least one inclusion keyword is required' });
    }

    // Create job
    const job = await SearchJob.create({
      userId: req.user!.id,
      name,
      keywordsInclude: keywords_include,
      keywordsExclude: keywords_exclude || [],
      semanticCriteria: semantic_criteria || null,
      semanticBatchMode: semantic_batch_mode !== undefined ? semantic_batch_mode : true,
      generateLatex: generate_latex !== undefined ? generate_latex : false,
      startYear: start_year,
      endYear: end_year,
      maxResults: max_results,
      status: 'pending',
    });

    // Add job to queue
    const jobId = await addSearchJobToQueue(job.id);
    job.celeryTaskId = jobId;
    await job.save();

    logger.info(`Search job created: ${job.id}`);

    return res.status(201).json(job);
  } catch (error) {
    logger.error('Error creating search job:', error);
    return res.status(500).json({ detail: 'Failed to create search job' });
  }
});

// GET /jobs - List all search jobs for current user
router.get('/', authenticate, async (req: AuthRequest, res: Response) => {
  try {
    const skip = parseInt(req.query.skip as string) || 0;
    const limit = parseInt(req.query.limit as string) || 100;

    const { count, rows } = await SearchJob.findAndCountAll({
      where: { userId: req.user!.id },
      order: [['createdAt', 'DESC']],
      offset: skip,
      limit: limit,
    });

    return res.json({ jobs: rows, total: count });
  } catch (error) {
    logger.error('Error listing search jobs:', error);
    return res.status(500).json({ detail: 'Failed to list search jobs' });
  }
});

// GET /jobs/:job_id - Get a specific search job
router.get('/:job_id', authenticate, async (req: AuthRequest, res: Response) => {
  try {
    const { job_id } = req.params;

    const job = await SearchJob.findOne({
      where: { id: job_id, userId: req.user!.id },
    });

    if (!job) {
      return res.status(404).json({ detail: 'Search job not found' });
    }

    return res.json(job);
  } catch (error) {
    logger.error('Error getting search job:', error);
    return res.status(500).json({ detail: 'Failed to get search job' });
  }
});

// PATCH /jobs/:job_id - Update a search job
router.patch('/:job_id', authenticate, async (req: AuthRequest, res: Response) => {
  try {
    const { job_id } = req.params;
    const { name, status } = req.body;

    const job = await SearchJob.findOne({
      where: { id: job_id, userId: req.user!.id },
    });

    if (!job) {
      return res.status(404).json({ detail: 'Search job not found' });
    }

    // Update fields
    if (name !== undefined) {
      job.name = name;
    }

    if (status !== undefined) {
      const validStatuses = ['pending', 'running', 'paused', 'completed', 'failed'];
      if (!validStatuses.includes(status)) {
        return res.status(400).json({ detail: `Invalid status. Must be one of: ${validStatuses.join(', ')}` });
      }
      job.status = status;
    }

    await job.save();

    return res.json(job);
  } catch (error) {
    logger.error('Error updating search job:', error);
    return res.status(500).json({ detail: 'Failed to update search job' });
  }
});

// DELETE /jobs/:job_id - Delete a search job
router.delete('/:job_id', authenticate, async (req: AuthRequest, res: Response) => {
  try {
    const { job_id } = req.params;

    const job = await SearchJob.findOne({
      where: { id: job_id, userId: req.user!.id },
    });

    if (!job) {
      return res.status(404).json({ detail: 'Search job not found' });
    }

    // Delete CSV file if exists
    if (job.csvFilePath && fs.existsSync(job.csvFilePath)) {
      fs.unlinkSync(job.csvFilePath);
    }

    await job.destroy();

    return res.status(204).send();
  } catch (error) {
    logger.error('Error deleting search job:', error);
    return res.status(500).json({ detail: 'Failed to delete search job' });
  }
});

// POST /jobs/:job_id/pause - Pause a running search job
router.post('/:job_id/pause', authenticate, async (req: AuthRequest, res: Response) => {
  try {
    const { job_id } = req.params;

    const job = await SearchJob.findOne({
      where: { id: job_id, userId: req.user!.id },
    });

    if (!job) {
      return res.status(404).json({ detail: 'Search job not found' });
    }

    if (job.status !== 'running') {
      return res.status(400).json({ detail: 'Only running jobs can be paused' });
    }

    job.status = 'paused';
    job.statusMessage = 'Pausing job...';
    await job.save();

    return res.json(job);
  } catch (error) {
    logger.error('Error pausing search job:', error);
    return res.status(500).json({ detail: 'Failed to pause search job' });
  }
});

// POST /jobs/:job_id/resume - Resume a failed or paused search job
router.post('/:job_id/resume', authenticate, async (req: AuthRequest, res: Response) => {
  try {
    const { job_id } = req.params;

    const job = await SearchJob.findOne({
      where: { id: job_id, userId: req.user!.id },
    });

    if (!job) {
      return res.status(404).json({ detail: 'Search job not found' });
    }

    if (!['failed', 'paused'].includes(job.status)) {
      return res.status(400).json({ detail: 'Only failed or paused jobs can be resumed' });
    }

    // Add job to queue for resumption
    await addSearchJobToQueue(job.id, true);

    return res.json(job);
  } catch (error) {
    logger.error('Error resuming search job:', error);
    return res.status(500).json({ detail: 'Failed to resume search job' });
  }
});

// GET /jobs/:job_id/papers - Get papers for a specific search job
router.get('/:job_id/papers', authenticate, async (req: AuthRequest, res: Response) => {
  try {
    const { job_id } = req.params;
    const skip = parseInt(req.query.skip as string) || 0;
    const limit = parseInt(req.query.limit as string) || 100;

    // Verify job belongs to current user
    const job = await SearchJob.findOne({
      where: { id: job_id, userId: req.user!.id },
    });

    if (!job) {
      return res.status(404).json({ detail: 'Search job not found' });
    }

    // Get papers for this job
    const { count, rows } = await Paper.findAndCountAll({
      where: { searchJobId: job_id },
      order: [['createdAt', 'DESC']],
      offset: skip,
      limit: limit,
    });

    return res.json({ papers: rows, total: count });
  } catch (error) {
    logger.error('Error getting papers:', error);
    return res.status(500).json({ detail: 'Failed to get papers' });
  }
});

// GET /jobs/:job_id/download - Download CSV results
router.get('/:job_id/download', authenticate, async (req: AuthRequest, res: Response) => {
  try {
    const { job_id } = req.params;

    const job = await SearchJob.findOne({
      where: { id: job_id, userId: req.user!.id },
    });

    if (!job) {
      return res.status(404).json({ detail: 'Search job not found' });
    }

    if (job.status !== 'completed') {
      return res.status(400).json({ detail: 'Job is not completed yet' });
    }

    if (!job.csvFilePath || !fs.existsSync(job.csvFilePath)) {
      return res.status(404).json({ detail: 'Results file not found' });
    }

    const filename = `${job.name.replace(/\s+/g, '_')}_${job.id}.csv`;

    return res.download(job.csvFilePath, filename);
  } catch (error) {
    logger.error('Error downloading results:', error);
    return res.status(500).json({ detail: 'Failed to download results' });
  }
});

// GET /jobs/:job_id/prisma-diagram - Download PRISMA diagram
router.get('/:job_id/prisma-diagram', authenticate, async (req: AuthRequest, res: Response) => {
  try {
    const { job_id } = req.params;

    const job = await SearchJob.findOne({
      where: { id: job_id, userId: req.user!.id },
    });

    if (!job) {
      return res.status(404).json({ detail: 'Search job not found' });
    }

    if (job.status !== 'completed') {
      return res.status(400).json({ detail: 'Job is not completed yet' });
    }

    if (!job.prismaDiagramPath || !fs.existsSync(job.prismaDiagramPath)) {
      return res.status(404).json({ detail: 'PRISMA diagram not found' });
    }

    const filename = `${job.name.replace(/\s+/g, '_')}_PRISMA_${job.id}.svg`;

    return res.download(job.prismaDiagramPath, filename);
  } catch (error) {
    logger.error('Error downloading PRISMA diagram:', error);
    return res.status(500).json({ detail: 'Failed to download PRISMA diagram' });
  }
});

// GET /jobs/:job_id/latex - Download LaTeX document
router.get('/:job_id/latex', authenticate, async (req: AuthRequest, res: Response) => {
  try {
    const { job_id } = req.params;

    const job = await SearchJob.findOne({
      where: { id: job_id, userId: req.user!.id },
    });

    if (!job) {
      return res.status(404).json({ detail: 'Search job not found' });
    }

    if (job.status !== 'completed') {
      return res.status(400).json({ detail: 'Job is not completed yet' });
    }

    if (!job.latexFilePath || !fs.existsSync(job.latexFilePath)) {
      return res.status(404).json({ detail: 'LaTeX document not found' });
    }

    const filename = `${job.name.replace(/\s+/g, '_')}_Review_${job.id}.tex`;

    return res.download(job.latexFilePath, filename);
  } catch (error) {
    logger.error('Error downloading LaTeX:', error);
    return res.status(500).json({ detail: 'Failed to download LaTeX document' });
  }
});

// GET /jobs/:job_id/bibtex - Download BibTeX file
router.get('/:job_id/bibtex', authenticate, async (req: AuthRequest, res: Response) => {
  try {
    const { job_id } = req.params;

    const job = await SearchJob.findOne({
      where: { id: job_id, userId: req.user!.id },
    });

    if (!job) {
      return res.status(404).json({ detail: 'Search job not found' });
    }

    if (job.status !== 'completed') {
      return res.status(400).json({ detail: 'Job is not completed yet' });
    }

    if (!job.bibtexFilePath || !fs.existsSync(job.bibtexFilePath)) {
      return res.status(404).json({ detail: 'BibTeX file not found' });
    }

    const filename = `${job.name.replace(/\s+/g, '_')}_References_${job.id}.bib`;

    return res.download(job.bibtexFilePath, filename);
  } catch (error) {
    logger.error('Error downloading BibTeX:', error);
    return res.status(500).json({ detail: 'Failed to download BibTeX file' });
  }
});

// GET /jobs/:job_id/screenshot - Get latest screenshot
router.get('/:job_id/screenshot', authenticate, async (req: AuthRequest, res: Response) => {
  try {
    const { job_id } = req.params;

    // Verify job belongs to current user
    const job = await SearchJob.findOne({
      where: { id: job_id, userId: req.user!.id },
    });

    if (!job) {
      return res.status(404).json({ detail: 'Search job not found' });
    }

    // Find the latest screenshot for this job
    const screenshotDir = path.join(config.UPLOAD_DIR, 'screenshots');
    const pattern = path.join(screenshotDir, `scraper_${job_id}_*.png`);

    const screenshots = await globAsync(pattern);

    if (!screenshots || screenshots.length === 0) {
      return res.status(404).json({ detail: 'No screenshot available yet' });
    }

    // Get the most recent screenshot
    const latestScreenshot = screenshots.sort((a: string, b: string) => {
      return fs.statSync(b).ctime.getTime() - fs.statSync(a).ctime.getTime();
    })[0];

    return res.sendFile(latestScreenshot, {
      headers: {
        'Cache-Control': 'no-cache, no-store, must-revalidate',
        Pragma: 'no-cache',
        Expires: '0',
      },
    });
  } catch (error) {
    logger.error('Error getting screenshot:', error);
    return res.status(500).json({ detail: 'Failed to get screenshot' });
  }
});

export default router;
